lib/bigshift/cli.rb

```diff
@@ -268,15 +278,27 @@ def aws_region
       @aws_region ||= ((awsc = @config[:aws_credentials]) && awsc['region']) || ENV['AWS_REGION'] || ENV['AWS_DEFAULT_REGION']
     end
 
-    def raw_gcp_credentials
-      @config[:gcp_credentials]
+    def gcp_project
+      if @config[:gcp_credentials]
+        @config[:gcp_credentials]['project_id']
+      else
+        Google::Cloud.env.project_id
+      end
     end
 
     def gcp_credentials
-      @gcp_credentials ||= Google::Auth::ServiceAccountCredentials.make_creds(
-        json_key_io: StringIO.new(JSON.dump(raw_gcp_credentials)),
-        scope: Google::Apis::StoragetransferV1::AUTH_CLOUD_PLATFORM
-      )
+      @gcp_credentials ||= begin
+        if @config[:gcp_credentials]
+          credentials = Google::Auth::ServiceAccountCredentials.make_creds(
+            json_key_io: StringIO.new(JSON.dump(@config[:gcp_credentials])),
+            scope: Google::Apis::StoragetransferV1::AUTH_CLOUD_PLATFORM
+          )
+        elsif (credentials = Google::Auth::GCECredentials.new)
+          credentials
+        else
+          raise 'No GCP credentials found'
```

Comment:
This should raise `CliError` like in other places in this file.

Reply:
I just saw I had missed this too when validating the AWS credentials. I don't remember if that was intentional or not. Either ignore the above comment, or change both instances.

Reply:
It turns out that in the absence of the credential files the other mechanism for both AWS and GCP will always create an object however you don't know if it's valid until you attempt to invoke the relevant API. After a bunch of testing I added some top level exception handling to tell the user their error in https://github.com/iconara/bigshift/blob/e0bc51be55652338231345edb7cada960e766f92/lib/bigshift/cli.rb#L27-L44

Reply:
This looks like a brittle solution, unfortunately. It's almost always wrong to catch `NoMethodError` (and "match" doesn't seem like a very specific filter).

Could you explain a bit more why this happens (the `NoMethodError`) and perhaps I can help you figure out how to make it less brittle.

Reply:
The error occurs if you don't specify the AWS credentials either in the environment or as a file. Google's SDK raises a specific exception but the AWS library sadly doesn't. I could also check the trace but that would make it even more brittle. I was trying to avoid writing an AWS configuration validation but perhaps that's the best thing to do? 

```AWS configuration missing or malformed: undefined method `match' for nil:NilClass:                                                                                                                                                                  
                                                                                                                                                                                                                                                    
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:82:in `block in partition_matching_region'                                                                           
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `each'                                                                                                         
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `find'                                                                                                         
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:81:in `partition_matching_region'                                                                                    
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:70:in `get_partition'                                                                                                
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:23:in `signing_region'                                                                                               
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-partitions-1.121.0/lib/aws-partitions/endpoint_provider.rb:99:in `signing_region'                                                                                               
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/plugins/s3_signer.rb:20:in `block in <class:S3Signer>'                                                                                             
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:70:in `call'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:211:in `block in resolve_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:57:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:57:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:210:in `resolve_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:203:in `value_at'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:189:in `block in resolve'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/2.5.0/set.rb:338:in `each_key'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/2.5.0/set.rb:338:in `each'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:189:in `resolve'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:177:in `apply_defaults'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/configuration.rb:150:in `build!'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:62:in `build_config'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:19:in `initialize'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/client.rb:262:in `initialize'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-core-3.42.0/lib/seahorse/client/base.rb:99:in `new'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/aws-sdk-s3-1.29.0/lib/aws-sdk-s3/resource.rb:14:in `initialize'
* /home/~/git/bigshift/lib/bigshift/cli.rb:218:in `new'
* /home/~/git/bigshift/lib/bigshift/cli.rb:218:in `s3_resource'
* /home/~/git/bigshift/lib/bigshift/cli.rb:229:in `create_unload_manifest'
* /home/~/git/bigshift/lib/bigshift/cli.rb:68:in `unload'
* /home/~/git/bigshift/lib/bigshift/cli.rb:30:in `run'
* /home/~/git/bigshift/bin/bigshift:6:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/bin/bigshift:23:in `load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/bin/bigshift:23:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:74:in `kernel_load'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli/exec.rb:28:in `run'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:463:in `exec'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/invocation.rb:126:in `invoke_command'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor.rb:387:in `dispatch'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:27:in `dispatch'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/vendor/thor/lib/thor/base.rb:466:in `start'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/cli.rb:18:in `start'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/exe/bundle:30:in `block in <top (required)>'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/lib/bundler/friendly_errors.rb:124:in `with_friendly_errors'
* /home/~/.rbenv/versions/2.5.3/lib/ruby/gems/2.5.0/gems/bundler-1.17.1/exe/bundle:22:in `<top (required)>'
* /home/~/.rbenv/versions/2.5.3/bin/bundle:23:in `load'
* /home/~/.rbenv/versions/2.5.3/bin/bundle:23:in `<main>'```